### PR TITLE
Software Processes

### PR DESCRIPTION
**Pull request by emartinez-usgs**
- Reorganized some parts of the repository
- Added software section
  - Attempted to provided implementation guidelines for various software related tasks

### CODE REVIEW

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Code of Conduct
+===============
+
+All contributions to- and interactions surrounding this project will abide by
```
Comment (jmfee-usgs) [inline CODE_OF_CONDUCT.md]: `to-`?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
-If you have changes you would like to contribute to the best practices
-guidelines, you can do that by forking the repository and creating a
-pull request back to this repository.
-- See: [Fork A Repo](https://help.github.com/articles/fork-a-repo/)
-- See: [Using pull requests](https://help.github.com/articles/using-pull-requests/)
+Contributing
+============
+
+Contributions are welcome from the community. Questions can be asked on the
+[issues page][1]. Before creating a new issue, please atake a moment to search
```
Comment (jmfee-usgs) [inline CONTRIBUTING.md]: `atake`?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
+and make sure a similar issue does not already exist. If one does exist, you
+can comment (most simply even with just a :+1:) to show your support for that
```
Comment (jmfee-usgs) [inline CONTRIBUTING.md]: escape `:+1:` in addition?

File: DISCLAIMER.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+Disclaimer
+==========
+
+Statements and processes described in this repository represent best practice
+guidance aimed at fostering an effective software development community within
+the U.S. Geological Survey (USGS). Official policy is maintained on the
+[U.S. Geological Survey website][1]. This repository is a living document,
```
Comment (jmfee-usgs) [inline DISCLAIMER.md]: Are you waiting for the link?

File: languages/README.md [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Languages
```
Comment (jmfee-usgs) [inline languages/README.md]: Should software languages be under the software directory?  Not sure if this repo would have information about other types of language (like English)

File: software/README.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+when collaboratively developing a project, however it also works when
+working as a solo developer on a project as well. This workflow encourages
+[pull requests][6] be used for integration. Pull requests facilitate peer review
+as you actively develop software.
+
+All software developed by the USGS should by default use an open-source license
+(put a "LICENSE.md in the project root") and should also include
+[appropriate preliminary (provisional) software disclaimers][7].
+
+See the [example project scaffold][8] for what a new project might look like.
```
Comment (jmfee-usgs) [inline software/README.md]: Is this the right link? `[8]` is referenced in the next paragraph and seems to be unrelated.

File: software/releases.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    potentially asking clarifying questions using the issue used to make the
+    request
+  - The _repository administrator_ may now import the software into the new
+    repository
+
+
+Preliminary Software Subsequent Release
+---------------------------------------
+
+Each commit made to the USGS organization repositroy should have received
```
Comment (jmfee-usgs) [inline software/releases.md]: `repositroy`

File: software/releases.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+   - You **must** mark the release as `pre-release`
+
+
+Approved Software Release
+-------------------------
+
+All approved software releases should be made from existing
+[preliminary software][2] repositories.
+
+ - Follow the steps for completing a _review for approved software_
```
Comment (jmfee-usgs) [inline software/releases.md]: Make this link to the appropriate `review for approved software` section?

File: software/reviews.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+   the primary USGS repository
+   - Assign (or request) a reviewer to the pull request
+ - Wait for reviewer to approve the changes
+   - This may involve iteratively fixing issues the reviewer points out
+ - Merge (or better, have the reviewer merge) the changes into the USGS
+   repository
+
+**For [approved software][6] tags...**
+
+ - Create a release candidate reconciliation _branch_
```
Comment (jmfee-usgs) [inline software/reviews.md]: Do you want to recommend a milestone to track any issues created during the review process?

File: software/stages.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+Scratch Software
+----------------
+
+These exist in folders on local workstations or possibly in the user account
+space on GitHub. For example: `https://github.com/{user}/{project}.git`. Where
+`{user}` is your user account name and `{project}` is the name of the software
+project repository.
+
+These are your own and no restrictions are placed on them.
```
Comment (jmfee-usgs) [inline software/stages.md]: Beyond not including sensitive information, right?

File: software/stages.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+These may be initiated directly in the USGS organization or may be
+migrated from a scratch software project. In the latter case, the scratch
+software project must receive a peer [administrative security review][1] prior
+to migration.
+
+
+Approved Software
+-----------------
+
+
```
Comment (jmfee-usgs) [inline software/stages.md]: Preliminary software that has gone through additional review and received approval?

File: software/types.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+below. There are different review requirements imposed on software depending
+on their software type categorization.
+
+
+Scientific Software
+-------------------
+
+Scientific software implements scientific techniques, methods and/or
+algorithms. Typically (but not always) these techniques, methods and/or
+algorithsms are documented in corresponding publications.
```
Comment (jmfee-usgs) [inline software/types.md]: `algorithsms`

File: software/types.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
+Software Types
+==============
+
+There are two primary categorizations for software types: (1) scientific
```
Comment (jmfee-usgs) [inline software/types.md]: Is there a hybrid, or if any scientific techniques, methods and/or algorithms are part of the software it is considered scientific?

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (emartinez-usgs) [inline CODE_OF_CONDUCT.md]: Fixed in: 1f7bc93

File: CONTRIBUTING.md (hunk 2 above)
Comment (emartinez-usgs) [inline CONTRIBUTING.md]: Fixed in: 9fbda2b

File: CONTRIBUTING.md (hunk 3 above)
Comment (emartinez-usgs) [inline CONTRIBUTING.md]: Fixed in: 68a303e

File: DISCLAIMER.md (hunk 4 above)
Comment (emartinez-usgs) [inline DISCLAIMER.md]: Fixed in: 8e0717f

File: languages/README.md (hunk 5 above)
Comment (emartinez-usgs) [inline languages/README.md]: Not sure either. We can move it later if we add information about wording, voice, (english) etc...

File: software/README.md (hunk 6 above)
Comment (emartinez-usgs) [inline software/README.md]: Fixed (ish) in: 5b3b1c0

This link doesn't exist until #41 is merged.

File: software/releases.md (hunk 7 above)
Comment (emartinez-usgs) [inline software/releases.md]: Fixed in: 83df21d

File: software/releases.md (hunk 8 above)
Comment (emartinez-usgs) [inline software/releases.md]: I agree in principle, but there is no such link at this time. Best I could do would be a link to the reviews page. Otherwise I might need to reformat the entire page.

File: software/reviews.md (hunk 9 above)
Comment (emartinez-usgs) [inline software/reviews.md]: Milestones present some potential logistical problems. They can work in certain cases as an alternative.

File: software/stages.md (hunk 10 above)
Comment (emartinez-usgs) [inline software/stages.md]: Depending on where they live and if they will ever be made public that is probably a good idea, but there are no hard restrictions restrictions on it.

File: software/stages.md (hunk 11 above)
Comment (emartinez-usgs) [inline software/stages.md]: Fixed in: f535320

File: software/types.md (hunk 13 above)
Comment (emartinez-usgs) [inline software/types.md]: For now, just `scientific` or `non-scientific`. As more types become well-defined we can update this documentation.

File: software/types.md (hunk 12 above)
Comment (emartinez-usgs) [inline software/types.md]: Fixed in: 611e6d2

**Comment by emartinez-usgs** [comment]
I've fixed or documented the `won't fix` for each of @jmfee-usgs recommendations.

File: DISCLAIMER.md [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+must defer to official USGS policy.
+
+This information is preliminary and is subject to revision. The information
+is provided on the condition that neither the U.S. Geological Survey nor
+the U.S. Government shall be held liable for any damages resulting from the
+authorized or unauthorized use of the information.
+
+
+
+[1]: https://www.usgs.gov/
```
Comment (mguy-usgs) [inline DISCLAIMER.md]: *maybe* this should be a reference actually to USGS Policy website pages, I say this because the USGS side can be difficult to navigate, and it seems to me are actually pointing people to usgs policies https://www2.usgs.gov/fsp/policies.asp

File: data/README.md [hunk 15]
```diff
@@ -0,0 +1,6 @@
+Data
+====
+
+This section contains best practices for data management.
```
Comment (mguy-usgs) [inline data/README.md]: Add in the same disclaimer you did about software, this is guidelines and the definitive instruction is USGS policy with link to USGS policy page.

File: data/README.md [hunk 16]
```diff
@@ -0,0 +1,6 @@
...
+
+> Guidance and best practices forthcoming...
```
Comment (mguy-usgs) [inline data/README.md]: I think it is best that we point to the USGS Data Management web pages here, again don't want to be redundant to where we are inconsistent with rest of USGS and there is significant information on this already at https://www2.usgs.gov/datamanagement/why.php

File: software/README.md [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+Your GitHub account is your own. You can [create public repositories][5] using
+your account. There is no limitation on what can be done in these repositories,
+however keep in mind that commits made to these repositories are public to the
+world. A best practice is to **never commit private or otherwise sensitive**
+information.
+
+While actively developing a project, it is recommended you follow an
+"[Integration Manager Workflow][5]". This workflow works especially well
+when collaboratively developing a project, however it also works when
```
Comment (mguy-usgs) [inline software/README.md]: add comma after however

File: software/README.md [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+working as a solo developer on a project as well. This workflow encourages
```
Comment (mguy-usgs) [inline software/README.md]: the "as well" is redundant with "also" in line 39, you only need one of these, see I am being redundant ;-)

File: software/README.md [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+[pull requests][6] be used for integration. Pull requests facilitate peer review
+as you actively develop software.
+
+All software developed by the USGS should by default use an open-source license
+(put a "LICENSE.md in the project root") and should also include
+[appropriate preliminary (provisional) software disclaimers][7].
```
Comment (mguy-usgs) [inline software/README.md]: yes, should have preliminary software disclaimer when developing under your github account, but it is not clear that all of this section "Start Developing" is ONLY about developing under your own github account. So maybe make that more clean in this section?

File: software/releases.md [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+   - This should include at least one [administrative security review][1],
+     one [code review][5], and two [subject matter expert reviews][6] (the
+     subject matter expert reviews only apply to [scientific software][7]
+     types)
+   - Note: A single person may complete more than one type of review,
+     completing the recommended review types may involve as few as two
+     reviewers
+ - Obtain science center director (or delegated authority) to publish an
```
Comment (mguy-usgs) [inline software/releases.md]: add in word "approval" so it reads: Obtain science center director (or delegated authority) approval to publish

File: software/reviews.md [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+slightly depending on the current project state.
+
+**For _new_ [preliminary software...][3]**
+
+> This review occurs prior to migrating scratch software into the USGS
+> organization on GitHub.
+
+ - Put the software in a version control repository accessible by the intended
+   reviewer
+ - Create an issue in the software release reviews project requesting
```
Comment (mguy-usgs) [inline software/reviews.md]: when this exists, can you put a link here?

File: software/stages.md [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+`{user}` is your user account name and `{project}` is the name of the software
+project repository.
+
+These are your own and no restrictions are placed on them.
+
+
+Preliminary Software
+--------------------
+
+These exist in the USGS organization on GitHub. For example:
```
Comment (mguy-usgs) [inline software/stages.md]: we all know github is publicly accessible, but I think it is good to state that again here to be clear

File: software/reviews.md [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+ - User input cleansing
+ - Memory leaks
+ - Vulnerabilities
+ - Optimizations
+ - ... etc ...
+
+These reviews should be performed by a **trusted developer** with applicable
+programming knowledge. This person need not have thorough subject-matter
+knowledge on the project topic.
+
```
Comment (llastowka-usgs) [inline software/reviews.md]: Can they request this review through GitHub channels?

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -1,11 +1,18 @@
...
+can comment (most simply even with just a `:+1:`) to show your support for that
+issue.
+
+If you have direct contributions you would like considered for incorporation
+into the project you can [fork this repository][2] and
+[submit a pull request][3] for review.
+
+
+
+[1]: ../issues/
```
Comment (emartinez-usgs) [inline CONTRIBUTING.md]: Fixed in: efbf47b

File: DISCLAIMER.md (hunk 14 above)
Comment (emartinez-usgs) [inline DISCLAIMER.md]: Fixed in: a9db54c

File: data/README.md (hunk 15 above)
Comment (emartinez-usgs) [inline data/README.md]: Hmm... I don't see that for software specifically. I think such a statement exists in the DISCLAIMER.md file for this repository as a whole as well as in the root README.md of the project-scaffold-generic branch.

We can probably think about adding such a disclaimer here at some future point if/when we actually populate some data guidelines.

File: data/README.md (hunk 16 above)
Comment (emartinez-usgs) [inline data/README.md]: Fixed in: 008e3ea

File: software/README.md (hunk 17 above)
Comment (emartinez-usgs) [inline software/README.md]: Fixed in: b4efc2d

File: software/README.md (hunk 18 above)
Comment (emartinez-usgs) [inline software/README.md]: Fixed in: b4efc2d

File: software/README.md (hunk 19 above)
Comment (emartinez-usgs) [inline software/README.md]: Fixed in: 729c438

File: software/releases.md (hunk 20 above)
Comment (emartinez-usgs) [inline software/releases.md]: Fixed in: 58e96b4

File: software/reviews.md (hunk 21 above)
Comment (emartinez-usgs) [inline software/reviews.md]: Fixed in: 78cfe0b

File: software/reviews.md (hunk 23 above)
Comment (emartinez-usgs) [inline software/reviews.md]: This section of the page only describes the type of review. The process for requesting and completing the review depends on the software's current [lifecycle] stage and is described earlier in this document.

File: software/stages.md (hunk 22 above)
Comment (emartinez-usgs) [inline software/stages.md]: Fixed in: 7780dcc

**Comment by emartinez-usgs** [comment]
Resolved or explained issues brought up by @mguy-usgs and @llastowka-usgs.
